addons/hr_timesheet/models/project.py

```diff
@@ -1,31 +1,99 @@
+    analytic_account_id = fields.Many2one('account.analytic.account', string="Analytic Account", ondelete='set null',
+        help="Link this project to an analytic account if you need financial management on projects. "
+             "It enables you to connect projects with budgets, planning, cost and revenue analysis, timesheets on projects, etc.")
+
+    @api.onchange('partner_id')
+    def _onchange_partner_id(self):
+        domain = []
+        if self.partner_id:
+            domain = [('partner_id', '=', self.partner_id.id)]
+        return {'domain': {'analytic_account_id': domain}}
+
+    @api.constrains('allow_timesheets', 'analytic_account_id')
+    def _check_allow_timesheet(self):
+        for project in self:
+            if project.allow_timesheets and not project.analytic_account_id:
+                raise ValidationError(_('To allow timesheet, your project %s should have an analytic account set.' % (project.name,)))
+
+    @api.model
+    def name_create(self, name):
+        """ Create a project with name_create should generate analytic account creation """
+        values = {
+            'name': name,
+            'allow_timesheets': True,
+        }
+        return self.create(values).name_get()[0]
+
+    @api.model
+    def create(self, values):
+        """ Create an analytic account if project allow timesheet and don't provide one """
+        if values.get('allow_timesheets') and not values.get('analytic_account_id'):
+            analytic_account = self._create_analytic_account(values)
+            values['analytic_account_id'] = analytic_account.id
+        return super(Project, self).create(values)
+
+    @api.multi
+    def write(self, values):
+        result = super(Project, self).write(values)
+        if values.get('allow_timesheets'):
+            for project in self:
+                if not project.analytic_account_id:
+                    analytic_account = project._create_analytic_account({
+                        'name': project.name,
+                        'company_id': project.company_id.id,
+                        'partner_id': project.partner_id.id,
+                    })
+                    project.write({'analytic_account_id': analytic_account.id})
+        return result
+
+    @api.multi
+    def unlink(self):
+        """ Delete the empty related analytic account """
+        analytic_accounts_to_delete = self.env['account.analytic.account']
+        for project in self:
+            if project.analytic_account_id and not project.analytic_account_id.line_ids:
+                analytic_accounts_to_delete |= project.analytic_account_id
+        result = super(Project, self).unlink()
+        analytic_accounts_to_delete.unlink()
+        return result
+
+    def _create_analytic_account(self, values):
+        analytic_account = self.env['account.analytic.account'].create({
+            'name': values.get('name', _('Unkwon Analytic Account')),
```

Comment:
[Commit4] Note sure: nothing about currency ? I suppose it comes from company and therefore not necessary to re-specify it ?

Reply:
currency field is a related on the company.

Reply:
Ok then.